lightpath/controller.py

```diff
@@ -165,18 +166,46 @@ def load_beamline(self, endstation: str):
             # Find the paths from each source to the desired line
             for src in self.sources:
                 try:
-                    if nx.has_path(self.graph, src, branch):
-                        found_paths = nx.all_simple_paths(self.graph,
-                                                          source=src,
-                                                          target=branch)
-                        paths.extend(found_paths)
-                    else:
-                        logger.debug(f'No path between {src} and {branch}')
+                    path_exists: bool = nx.has_path(self.graph, src, branch)
                 except NodeNotFound:
                     logger.debug(f'Either source {src} or target {branch} '
                                  'not found.')
+                    path_exists = False
+
+                if path_exists:
+                    found_paths = nx.all_simple_paths(self.graph,
+                                                      source=src,
+                                                      target=branch)
+
+                    # Trim beamline based on z-positions
+                    if isinstance(end_branches, dict):
+                        new_paths = list()
+                        for found_path in found_paths:
+                            end_z = end_branches[branch]
+                            if end_z is not None:
+                                # Only filter internal devices
```

Comment:
What's the rationale or only filtering the internal devices? Are the first and last path elements something special?

Reply:
The first and last nodes of each path are special no-device nodes we use as end points, typically something like `source_L0` or `L2` or `K0`.  These have no device metadata associated with them (or no metadata at all, really)